mlflow/sagemaker/__init__.py

```diff
@@ -2597,14 +2598,17 @@ def predict(self, deployment_name=None, df=None, endpoint=None):
             sage_client = boto3.client(
                 "sagemaker-runtime", region_name=self.region_name, **assume_role_credentials
             )
+            if isinstance(df, pd.DataFrame):
+                body = (json.dumps({"dataframe_split": df.to_dict(orient="split")}),)
+            else:
+                body = json.dumps({"instances": _get_jsonable_obj(df)})
             response = sage_client.invoke_endpoint(
                 EndpointName=deployment_name,
-                Body=json.dumps(_get_jsonable_obj(df, pandas_orient="split")),
+                Body=body,
                 ContentType="application/json",
             )
-
             response_body = response["Body"].read().decode("utf-8")
-            return infer_and_parse_json_input(response_body)
+            return ScoringServerResponse.from_raw_json(response_body)
```

Comment:
This breaks the API contract of `BaseDeploymenntClient.predict()`: https://mlflow.org/docs/latest/python_api/mlflow.deployments.html#mlflow.deployments.BaseDeploymentClient.predict, which states that the API returns a pandas DataFrame, series, or numpy array (i.e pyfunc output, though list should also be added to the docs)

The goal of the original implementation was to convert the response JSON into the corresponding output object. Can we preserve that?

Reply:
Ah, this is a hard one. We could certainly preserve the behavior but I am not sure if we should because:
1. we wrap the response in a dictionary so we can return additional stuff. I think we should enable access to any additional output via the client interface.
2. It's actually impossible to know whether the response should be pandas dataframe or series or numpy array. I saw most clients will always return pandas dataframe. infer_and_parse_json returns a dataframe unless the input is explicitly a tensor input - i.e. a dict with `instances` or `inputs` field. Unfortunately, our current server does not make that distinction and just jsonizes whatever the model returned. It is a bit late to change that now unfortunately. 

Therefore, I would propose we update the interface to return pyfunc reponse object that can access the prediction as dataframe or numpy array per user's request and will have access to all response fields. Wdyt? Is that too big a change too late? We could also preserve the original behavior by default and add a flag to use the new behavior.